sklearn/linear_model/sag.py

```diff
@@ -0,0 +1,327 @@
+
+    Returns
+    -------
+    coef_init : array, shape = [n_features]
+        Weight vector.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import linear_model
+    >>> n_samples, n_features = 10, 5
+    >>> np.random.seed(0)
+    >>> y = np.random.randn(n_samples)
+    >>> X = np.random.randn(n_samples, n_features)
+    >>> clf = linear_model.Ridge(solver='sag')
+    >>> clf.fit(X, y)
+    ... #doctest: +NORMALIZE_WHITESPACE
+    Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
+          normalize=False, random_state=None, solver='sag', tol=0.001)
+
+    See also
+    --------
+    Ridge, SGDRegressor, ElasticNet, Lasso, SVR
+    """
+    if max_iter is None:
+        max_iter = 1000
+
+    n_samples, n_features = X.shape[0], X.shape[1]
+    alpha = float(alpha) / n_samples
+    fit_intercept = False
+
+    # initialization
+    if sample_weight is None:
+        sample_weight = np.ones(n_samples, dtype=np.float64, order='C')
+    coef_init = np.zeros(n_features, dtype=np.float64, order='C')
+    intercept_init = 0.0
+    weight_pos = 1.0
+    weight_neg = 1.0
+
+    # TODO: *_init (with a boolean warm-start) as parameters ?
+    intercept_sum_gradient_init = 0.0
+    sum_gradient_init = np.zeros(n_features, dtype=np.float64, order='C')
+    gradient_memory_init = np.zeros(n_samples, dtype=np.float64, order='C')
+    seen_init = np.zeros(n_samples, dtype=np.int32, order='C')
+    num_seen_init = 0
+
+    dataset, intercept_decay = make_dataset(X, y, sample_weight, random_state)
+
+    # set the eta0 at 1 / 4L where L is the max sum of
+    # squares for over all samples
+    step_size = get_auto_eta(dataset, alpha, n_samples, SquaredLoss(),
+                             fit_intercept)
+    if step_size * alpha == 1:
+        raise ZeroDivisionError("Current sag implementation does not handle "
+                                "the case step_size * alpha == 1")
+
+    intercept_, num_seen, max_iter_reached, intercept_sum_gradient = \
+        sag_sparse(dataset, coef_init.ravel(),
+                   intercept_init, n_samples,
+                   n_features, tol,
+                   max_iter,
+                   SquaredLoss(),
+                   step_size, alpha,
+                   sum_gradient_init.ravel(),
+                   gradient_memory_init.ravel(),
+                   seen_init.ravel(),
+                   num_seen_init,
+                   weight_pos, weight_neg,
+                   fit_intercept,
+                   intercept_sum_gradient_init,
+                   intercept_decay,
+                   verbose)
+
+    if max_iter_reached:
+        warnings.warn("The max_iter was reached which means "
+                      "the coef_ did not converge", ConvergenceWarning)
+
+    return coef_init
+
+
+def sag_logistic(X, y, coef_init, alpha=1e-4, sample_weight=None,
+                 max_iter=1000, tol=0.001, verbose=0, random_state=None,
+                 warm_start_mem=None):
```

Comment:
the init and warm start params should be at the end. For example here logistic and ridge SAG have different API.

also why 2 warm start param? We should allow to have only 1 passed last.